유저 프로필 추가 및 기존 API에 유저 이미지 url 추가

yooooonshine opened the pull request:
## 개요
- close #97

## 작업사항
- 유저 프로필 생성, 삭제 기능 추가


### User 안에 userImageUrl 추가
UserImageEntity로 url을 가져올 수 있으나, 매번 UserImage를 추가로 조회하는 것은 쿼리 비용이 커
User 안에 UserImageUrl 필드를 추가했습니다.

### ImageService 삭제
ImageService의 모든 메서드가 다른 로직에서 필요하지 않는 경우도 있으며, 파라미터가 일치하지 않는경우도 있어 아예 삭제했습니다.


### 포스트 리뷰 댓글 목록 조회 쿼리 수정
```
	@Query("""
			SELECT new hanium.modic.backend.web.postReview.dto.response.PostReviewCommentResponse(
				c.userId,
				u.name,
				c.createAt,
				c.text,
				CASE WHEN u.userImageUrl IS NOT NULL THEN true ELSE false END,
				u.userImageUrl
			)
			FROM PostReviewCommentEntity c
			JOIN UserEntity u ON c.userId = u.id
			WHERE c.postReviewId = :postReviewId
			ORDER BY c.createAt DESC
		""")
```
포스트 리뷰 댓글 목록 조회는
1. 목록 조회
2. 각 목록 엔티티에 따른 User, 조회
등으로 N + 1 문제가 발생하게 되어, 
아예 Join 쿼리로 한번에 가져오게 했습니다.



## 도움요청(해결)
```
	// 포스트 리뷰 목록 조회
	// TODO : 쿼리 최적화 필요.
	public Page<PostReviewDetailResponse> getPostReviews(final Long postId, final int page, final int size) {
		return postReviewRepository.findAllByPostId(postId, PageRequest.of(page, size))
			.map(postReview -> {
				// 회원 탈퇴 시 soft 탈퇴이므로 회원은 반드시 존재
				UserEntity user = userEntityRepository.findById(postReview.getUserId())
					.orElseThrow(() -> new AppException(USER_NOT_FOUND_EXCEPTION));

				final String userName = user.getName();
				final String userImageUrl = user.getUserImageUrl();
				final boolean hasImage = userImageUrl != null;
				final List<String> imageUrls = postReviewImageRepository.findAllByPostReviewId(postReview.getId())
					.stream()
					.map(postReviewImage -> postReviewImageService.createImageGetUrl(postReviewImage.getId()))
					.toList();

				return new PostReviewDetailResponse(
					userName,
					hasImage,
					userImageUrl,
					postReview.getCreateAt(),
					postReview.getId(),
					postReview.getDescription(),
					imageUrls
				);
			});
	}

```

현재 이 메서드의 경우
1. 포스트 리뷰 목록 요청
2. 각 포스트 리뷰당 user 조회, Image 목록 조회
쿼리가 나갑니다.

이로 인해
1 + 20 = 21 번의 쿼리가 나가게 됩니다.

이를 3중 join으로 처리할까하다가, 성능이 너무 저하될 거라 판단해 내비뒀습니다...

혹시 좋은 방법 없을까요 ㅠ

### 해결
![image](https://github.com/user-attachments/assets/67d59d34-b80d-4617-8625-24853185f412)
위와 같이 연관 데이터는 in쿼리를 통해 조회해오도록 했습니다!

또한 in쿼리 최적화를 위해 관련 필드에 대해서는 인덱스를 걸어뒀습니다!


이렇게 수정하고 실제 쿼리를 보니


```
Hibernate: 
    select
        pre1_0.id,
        pre1_0.create_at,
        pre1_0.description,
        pre1_0.post_id,
        pre1_0.update_at,
        pre1_0.user_id 
    from
        post_reviews pre1_0 
    where
        pre1_0.post_id=? 
    limit
        ?
Hibernate: 
    select
        count(pre1_0.id) 
    from
        post_reviews pre1_0 
    where
        pre1_0.post_id=?
Hibernate: 
    select
        ue1_0.id,
        ue1_0.coin,
        ue1_0.create_at,
        ue1_0.email,
        ue1_0.name,
        ue1_0.password,
        ue1_0.unique_id,
        ue1_0.update_at,
        ue1_0.user_image_url,
        ue1_0.user_role 
    from
        users ue1_0 
    where
        ue1_0.id in (?)
Hibernate: 
    select
        prie1_0.id,
        prie1_0.create_at,
        prie1_0.extension,
        prie1_0.full_image_name,
        prie1_0.image_name,
        prie1_0.image_path,
        prie1_0.image_purpose,
        prie1_0.image_url,
        prie1_0.post_review_id,
        prie1_0.update_at 
    from
        post_review_image prie1_0 
    where
        prie1_0.post_review_id in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
```

위와 같이 2번의 in쿼리로 데이터를 조회해옴을 확인했습니다.
그리고 MySQL 내부적으로 이 쿼리가 어떻게 동작하는 지 살펴봤습니다.

대략 2만 건의 더미데이터를 넣어 쿼리 동작을 확인했고
![image](https://github.com/user-attachments/assets/34ba8d59-a8e2-4262-bae7-e8a445bd27af)
확인 결과 인덱스 레인지 스캔이 발생함을 알 수 있었습니다.

이에 대해 정리한 글을 아래 첨부합니다!
https://mellow-licorice-caf.notion.site/20-in-2299e3adf6b4806e861bfc3a113abfcd?source=copy_link

yooooonshine committed:
refactor: ImageService에 saveImage 제거

yooooonshine committed:
refactor: ImageService 추상 클래스 제거
- ImageService 구현체마다 필요한 파라미터가 달라서 공통적으로 적용하기 어려웠다. 따라서 제거했다.

yooooonshine committed:
feat: 유저 Image 로직 구현

yooooonshine committed:
feat: 기존 API에 유저 이미지 필요 시 함께 반환하도록 변경

goalSetter09 commented on src/main/java/hanium/modic/backend/domain/user/entity/UserEntity.java:
```diff
@@ -44,6 +44,8 @@ public class UserEntity extends BaseEntity {
 
 	private Long coin = 0L;
 
+	private String userImageUrl;
```
여기도 다른 필드들 처럼 명시적으로 `@Column` 붙이는 것 어떤가용?

goalSetter09 replied:
그리고 그러면 `userImageEntity` 테이블은 따로 존재하고, 이 필드는 추가적으로 역정규화라고 생각하면 되나용?

yooooonshine replied:
넵 맞습니다!

혹시 `@Column`을 붙이는 게 좋다고 생각하는 이유를 여쭤볼 수 있을까용?

goalSetter09 replied:
필드명이 바뀌어도 db 컬럼명이 자동으로 바뀌지 않아서 안전하다..? 이런 이유로 붙이는게 좋다고 하더라구요..?!

yooooonshine replied:
오호 그렇군요 좋은 생각이네요!! 이것도 참고해서 수정하겠습니다!

goalSetter09 commented on src/main/java/hanium/modic/backend/web/follow/dto/response/GetFollowingsResponse.java:
```diff
@@ -1,7 +1,14 @@
+@JsonInclude(NON_NULL)
 public record GetFollowingsResponse(
 	Long userId,
+	boolean userImageExists,
```
`hasUserImage` 네이밍 가져가는 것 어떤가용?

boolean 필드의 경우에 롬복에서도 자동으로 getter 생성시 `has~`로 생성되기도 하고 좀더 자연스럽기도 하다고 생각합니다~!

yooooonshine replied:
오 좋습니다 수정하겠습니다!

goalSetter09 commented on src/main/java/hanium/modic/backend/web/profile/dto/GetProfileResponse.java:
```diff
@@ -1,8 +1,14 @@
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+@JsonInclude(NON_NULL)
```
여기서 이 어노테이션이 있는 이유가 profileImageUrl이 null인 경우 아예 안 보이게 하는 용도인가요??
저는 null이라도 보내는 게 맞다고 생각했는데 어떻게 생각하시는지 궁금합니다!

이유는 null일 때 보내지 않으면 응답 스키마 자체가 달라져서 프론트 측에서 처리할 때 어려움이 있을 수 있을 것 같아요!

그리고 

`boolean userImageExists,`
없이 그냥 `profileImageUrl`이 없으면 null로 보내는 방법은 별로인가요?
이건 그냥 궁금합니다..!

yooooonshine replied:
오 저는 프론트 측에 처리 생각해서 이렇게 한거긴 합니다!
단순히 프론트 측에서 null 체크를 해서 처리하는 것보다 bool값을 통해 true일 때만 접근하도록 하게 하는 게 목적이었어요!

이 부분은 프론트 분께 어떤게 편한지 여쭤보고 수정해보겠습니다!

goalSetter09 commented on src/main/java/hanium/modic/backend/domain/postReview/repository/PostReviewCommentRepository.java:
```diff
@@ -3,12 +3,29 @@
+			WHERE c.postReviewId = :postReviewId
+			ORDER BY c.createAt DESC
+		""")
+	Page<PostReviewCommentResponse> findAllByPostReviewIdOrderByCreateAtDesc(Long postReviewId, Pageable pageable);
```
`PostReviewCommentResponse`가 respository까지 넘겨지는게 좀 걱정되긴 하네용..
근데 이미 service에서도 쓰고 있긴 해서 별 문제 없을지도..?!

yooooonshine replied:
그춍..사실 역류이긴하죠 ㅎ..
controller 단에서 써야할 ResponseDto를 이미 Service에서 쓰고 있어서, 일단 진행하긴 헀는데
만약 역류로 인해 추후 불편한 점이 있으면 다시 고려해봐야 할 거 같긴 합니다!

goalSetter09 reviewed (commented):
고생하셨습니다!!!

image 로직은 조금 더 고민해보고 좋은 방법이 있는지 찾아볼게용

goalSetter09 commented:
리뷰와 유저는 조인으로 가져오고,
reviewId를 리스트로 담아서, IN 쿼리를 통해서 해당되는 이미지를 한 번에 가져와서 매핑시키는 방식은 어떨까요?

쿼리 두방에 가져올 수 있을 것 같아용

```
  Map<Long, List<PostReviewImageEntity>> imageMap = postReviewImageRepository
      .findAllByPostReviewIds(reviewIds)
      .stream()
      .collect(Collectors.groupingBy(PostReviewImageEntity::getPostReviewId));
```

이런 느낌..?

yooooonshine committed:
fix: createImageGetUrl 메서드 오류 수정
- userId 대신 imageId 들어가는 오류 수정

yooooonshine committed:
refactor: imageExists 변수면 has~~로 변경

yooooonshine committed:
refactor: getPostReviews 쿼리 최적화

- N + 1 문제를 in 쿼리로 해결

yooooonshine committed:
refactor: 엔티티에 필드명 명시 추가

goalSetter09 commented:
/gemini review

yooooonshine commented:
/gemini review

yooooonshine commented on src/main/java/hanium/modic/backend/domain/user/entity/UserImageEntity.java:
```diff
@@ -0,0 +1,41 @@
+package hanium.modic.backend.domain.user.entity;
+
+import org.springframework.stereotype.Indexed;
```
삭제했어!

yooooonshine committed:
fix: 트랜잭션 처리를 안해서 createImageGetUrl 메서드 요청시 다시 쿼리를 날리는 문제 수정
- getPostReviews메서드에서 PostReviewImage를 in쿼리로 조회해 createImageGetUrl메서드에서는 캐시되어 다시 쿼리를 안날릴 줄 알았으나 트랜잭션이 적용되지 않아 캐시가 안되는 문제점을 파악해 수정

yooooonshine committed:
refactor: Entity에 null 조건 추가함에 따른 테스트 코드 변경
